python/mxnet/gluon/probability/transformation/transformation.py

```diff
@@ -0,0 +1,289 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+"""Transformation Classes"""
+__all__ = ["Transformation", "TransformBlock","ComposeTransform", "ExpTransform",
+           "AffineTransform", "PowerTransform", "AbsTransform", 'SigmoidTransform',
+           'SoftmaxTransform']
+
+from ..distributions.utils import _clip_prob, cached_property, sum_right_most
+from ...block import HybridBlock
+import weakref
+
+
+class Transformation(object):
+    r"""Abstract class for implementing invertible transformation
+    with computable log  det jacobians
+    
+    Attributes
+    ----------
+    bijective : bool
+        
+    """
+    bijective = False
+    event_dim = 0
+
+    def __init__(self, F=None):
+        self._inv = None
+        self._F = F
+        super(Transformation, self).__init__()
+
+    @property
+    def F(self):
+        return self._F
+
+    @F.setter
+    def F(self, value):
+        self._F = value
+
+    @property
+    def sign(self):
+        """
+        Returns the sign of the determinant of the Jacobian.
+        """
+        raise NotImplementedError
+
+    @property
+    def inv(self):
+        inv = None
+        if self._inv is not None:
+            inv = self._inv()
+        if inv is None:
+            inv = _InverseTransformation(self)
+            self._inv = weakref.ref(inv)
```

Comment:
The idea is to cache the inverse transformation as long as it's in use somewhere else?

Reply:
yes